libvast/src/transform.cpp

```diff
@@ -109,57 +111,77 @@ transformation_engine::transformation_engine(std::vector<transform>&& transforms
 }
 
 /// Apply relevant transformations to the table slice.
-caf::expected<table_slice> transformation_engine::apply(table_slice&& x) const {
-  auto offset = x.offset();
-  auto size = x.rows();
-  auto layout = x.layout();
-  // TODO: Consider using a tsl robin map instead for transparent key lookup.
-  const auto& matching = layout_mapping_.find(std::string{layout.name()});
-  if (matching == layout_mapping_.end())
-    return std::move(x);
-  const auto& indices = matching->second;
-  VAST_INFO("applying {} transforms for received table slice w/ layout {}",
-            indices.size(), x.layout());
-  auto arrow_fast_path
-    = std::all_of(indices.begin(), indices.end(), [&](size_t idx) {
-        return transforms_.at(idx).arrow_fast_path_;
-      });
-  if (arrow_fast_path) {
-    VAST_INFO("selected fast path because all transforms support arrow");
-    // NOTE: It's important that `batch` is kept alive until `create()`
-    // is finished: If a copy was made, `batch` will hold the only reference
-    // to its underlying table slice, but the RecordBatches created by the
-    // transform step will most likely reference some of same underlying data.
-    auto original_batch = as_record_batch(x);
-    auto batch = original_batch;
+caf::error transformation_engine::add(table_slice&& x) {
+  auto batch = as_record_batch(x);
+  to_transform_[x.layout()].emplace_back(x.layout(), std::move(batch));
+  return caf::none;
+}
+
+caf::error
+transformation_engine::process_queue(transform& transform, batch_queue& queue) {
+  caf::error failed{};
+  const auto size = queue.size();
+  for (size_t i = 0; i < size; ++i) {
+    auto x = std::move(queue.front());
+    queue.pop_front();
+    auto add_failed = transform.add(x.first, x.second);
+    if (add_failed) {
+      failed = add_failed;
+      while (!queue.empty())
+        queue.pop_front();
+      break;
+    }
+  }
+  // Finish frees up resource inside the plugin.
+  auto finished = transform.finish();
+  if (!finished && !failed)
+    failed = std::move(finished.error());
+  if (failed)
+    return failed;
+  for (const auto& b : *finished)
+    queue.push_back(b);
+  return caf::none;
+}
+
+/// Apply relevant transformations to the table slice.
+caf::expected<std::vector<table_slice>> transformation_engine::finish() {
+  // NOTE: It's important that `batch` is kept alive until `create()`
+  // is finished: If a copy was made, `batch` will hold the only reference
+  // to its underlying table slice, but the RecordBatches created by the
+  // transform step will most likely reference some of same underlying data.
+  std::vector<std::shared_ptr<arrow::RecordBatch>> keep_alive{};
```

Comment:
@dominiklohmann, @lava : Is this keep_alive variable useful and sufficient in this case?

Reply:
I don't think we need this, actually.

Given `auto batch = as_record_batch(slice)` the deleter of `batch` should hold reference to both `slice` and the actually underlying record batch of `slice`, such that the lifetimes between `batch` and `slice` are properly shared. The existence of the batch thus implies that its underlying data is still present.

Reply:
Did this change in the meantime since we implemented transforms?

Reply:
Aside from a bug fix, no, I don't think so.

Reply:
Ok, I was asking because we added the comment and implemented the keepalive workaround after observing actual segfaults when using arrow transforms. So I don't really understand the difference, ie. why it was necessary in the old version but not anymore after the rewrite.

Reply:
Do you still have the context for that segfault? I can't imagine why this would've caused it.

Reply:
The `keep_alive` vector has a beneficial effect. The integration test will fail when the following lines are commented out:

```
    for (auto& [rows, offset, layout, batch] : bq)
      keep_alive.push_back(batch);
```

integration test:

```
sh build/vast/integration-Debug.sh -t transforms$ -v DEBUG
```

Reply:
But _why_ do we need it? There's a conflict between the actual reason and my mental model of the ownership, and I worry that we have a bug that's somewhere deeper in VAST that we yet need to understand.

Reply:
As far as I understand it's exactly for the reason that's described in the comment: When we create a new arrow record batch using `as_record_batch(slice)`, this slice will extend the lifetime of the underlying table slice, with which it may share some memory.

```
// table_slice.cpp
        auto result = std::shared_ptr<arrow::RecordBatch>{
          data,
          [batch = std::move(batch), slice](arrow::RecordBatch*) noexcept {
            static_cast<void>(batch);
            static_cast<void>(slice);
          }};
```

When during some transformation the batch is discarded and a new one is created as result, the table slice loses its last reference count, gets destroyed, and the batch (which still may share some of the memory) points into an unmapped memory region.

Reply:
So let me summarize my understanding:

We have table slice A and record batch B created using `B = as_record_batch(A)`. B references memory owned by A; A and B extend each other's lifetime so the memory is only free'd when both the reference count of A and B equal 0.

The issue occurs when we create a new record batch C that structurally shares memory with B. If A and B drop their reference count to 0 then C may point into free'd memory.

There are two possible solutions here:
1. Manually ensure that A or B is kept alive as long as C is alive.
2. Tie the lifetime of C to either A or B.

We're currently doing (1). That means that if we are ever to return C to an outside context we need to manually ensure that A or B are still alive as long as we use C.

I propose that we switch to doing (2) instead, which causes C's refcount to actually be meaningful. Owning C should imply that its underlying memory is still alive. Imagine what happens if a plugin developer stores C for later use: the obvious assumption would be that this is a valid thing to do, but it most certainly is not. I think the lifetime extension can easily be implemented when invoking the transform step so the plugin developer does not need to know about this.

Reply:
I think it's a good idea to think about how to solve the issue, but it's orthogonal to this PR. The issue is not new, so it's not a regression either.

That said, I also don't really see how to solve this on our side. We'd either need a  `RecordBatch::Copy()` function that we can call in `as_record_batch()` (but that may have severe performance implications), or ideally arrow needs to ensure that record batches which share memory also always share the same destructor. (actually this might be an upstream bug, now that I think about it)

Reply:
I agree that it shouldn't be a blocker for this PR, but we should definitely switch from (1) to (2) at some point in the future and we need a separate story for it so we actually do it. We should design the API to be as friendly to the plugin developer as possible.

Reply:
I've created https://app.shortcut.com/tenzir/story/31007/implement-proper-lifetime-extension-for-record-batches-created-from-arrow-table-slices